File: android/src/main/java/com/eliazavatta/plugins/metaads/MetaAds.java
package com.eliazavatta.plugins.metaads;

import android.content.Context;
import android.util.Log;
import com.facebook.ads.Ad;
import com.facebook.ads.AdError;
import com.facebook.ads.AdSettings;
import com.facebook.ads.AudienceNetworkAds;
import com.facebook.ads.InterstitialAd;
import com.facebook.ads.InterstitialAdListener;
import com.facebook.ads.RewardedVideoAd;
import com.facebook.ads.RewardedVideoAdListener;
import java.util.ArrayList;
import java.util.List;

public class MetaAds {

    private static final String TAG = "MetaAds";

    private RewardedVideoAd rewardedVideoAd;
    private InterstitialAd interstitialAd;
    private boolean isInitialized = false;
    private List<String> testDevices = new ArrayList<>();

    // Store callbacks for show events
    private RewardedVideoCallback currentRewardedCallback;
    private InterstitialCallback currentInterstitialCallback;

    // Callback interfaces
    public interface InitializationCallback {
        void onSuccess();
        void onError(String error);
    }

    public interface AdLoadCallback {
        void onAdLoaded();
        void onError(String error);
    }

    public interface RewardedVideoCallback {
        void onAdShown();
        void onRewardEarned(String rewardType, int rewardAmount);
        void onAdClosed();
        void onError(String error);
    }

    public interface InterstitialCallback {
        void onAdShown();
        void onAdClosed();
        void onError(String error);
    }

    public void initialize(Context context, String appId, boolean testMode, InitializationCallback callback) {
        try {
            Log.d(TAG, "Initializing Meta Audience Network with App ID: " + appId);

            // Set test mode if enabled
            if (testMode) {
                AdSettings.setTestMode(true);
                Log.d(TAG, "Test mode enabled");
            }

            // Add test devices if any
            if (!testDevices.isEmpty()) {
                AdSettings.addTestDevices(testDevices);
                Log.d(TAG, "Added " + testDevices.size() + " test devices");
            }

            // Initialize Audience Network
            AudienceNetworkAds.initialize(context);

            isInitialized = true;
            Log.d(TAG, "Meta Audience Network initialized successfully");
            callback.onSuccess();
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Meta Audience Network", e);
            callback.onError("Initialization failed: " + e.getMessage());
        }
    }

    public void loadRewardedVideo(Context context, String placementId, AdLoadCallback callback) {
        if (!isInitialized) {
            callback.onError("Meta Audience Network not initialized");
            return;
        }

        try {
            Log.d(TAG, "Loading rewarded video with placement ID: " + placementId);

            rewardedVideoAd = new RewardedVideoAd(context, placementId);

            RewardedVideoAdListener rewardedVideoAdListener = new RewardedVideoAdListener() {
                @Override
                public void onError(Ad ad, AdError adError) {
                    Log.e(TAG, "Rewarded video error: " + adError.getErrorMessage());
                    // Handle both load and show errors
                    if (currentRewardedCallback != null) {
                        currentRewardedCallback.onError("Error with rewarded video: " + adError.getErrorMessage());
                        currentRewardedCallback = null;
                    } else {
                        callback.onError("Failed to load rewarded video: " + adError.getErrorMessage());
                    }
                }

                @Override
                public void onAdLoaded(Ad ad) {
                    Log.d(TAG, "Rewarded video loaded successfully");
                    callback.onAdLoaded();
                }

                @Override
                public void onAdClicked(Ad ad) {
                    Log.d(TAG, "Rewarded video clicked");
                }

                @Override
                public void onLoggingImpression(Ad ad) {
                    Log.d(TAG, "Rewarded video impression logged");
                    if (currentRewardedCallback != null) {
                        currentRewardedCallback.onAdShown();
                    }
                }

                @Override
                public void onRewardedVideoCompleted() {
                    Log.d(TAG, "Rewarded video completed - reward earned");
                    if (currentRewardedCallback != null) {
                        currentRewardedCallback.onRewardEarned("coins", 1);
                        currentRewardedCallback = null;
                    }
                }

                @Override
                public void onRewardedVideoClosed() {
                    Log.d(TAG, "Rewarded video closed");
                    if (currentRewardedCallback != null) {
                        currentRewardedCallback.onAdClosed();
                        currentRewardedCallback = null;
                    }
                }
            };

            rewardedVideoAd.loadAd(rewardedVideoAd.buildLoadAdConfig().withAdListener(rewardedVideoAdListener).build());
        } catch (Exception e) {
            Log.e(TAG, "Error loading rewarded video", e);
            callback.onError("Error loading rewarded video: " + e.getMessage());
        }
    }

    public void showRewardedVideo(RewardedVideoCallback callback) {
        if (rewardedVideoAd == null || !rewardedVideoAd.isAdLoaded()) {
            callback.onError("Rewarded video not loaded");
            return;
        }

        try {
            Log.d(TAG, "Showing rewarded video");

            // Store callback for the unified listener to use
            currentRewardedCallback = callback;

            rewardedVideoAd.show();
        } catch (Exception e) {
            Log.e(TAG, "Error showing rewarded video", e);
            currentRewardedCallback = null;
            callback.onError("Error showing rewarded video: " + e.getMessage());
        }
    }

    public boolean isRewardedVideoLoaded() {
        return rewardedVideoAd != null && rewardedVideoAd.isAdLoaded();
    }

    public void loadInterstitial(Context context, String placementId, AdLoadCallback callback) {
        if (!isInitialized) {
            callback.onError("Meta Audience Network not initialized");
            return;
        }

        try {
            Log.d(TAG, "Loading interstitial with placement ID: " + placementId);

            interstitialAd = new InterstitialAd(context, placementId);

            InterstitialAdListener interstitialAdListener = new InterstitialAdListener() {
                @Override
                public void onError(Ad ad, AdError adError) {
                    Log.e(TAG, "Interstitial error: " + adError.getErrorMessage());
                    // Handle both load and show errors
                    if (currentInterstitialCallback != null) {
                        currentInterstitialCallback.onError("Error with interstitial: " + adError.getErrorMessage());
                        currentInterstitialCallback = null;
                    } else {
                        callback.onError("Failed to load interstitial: " + adError.getErrorMessage());
                    }
                }

                @Override
                public void onAdLoaded(Ad ad) {
                    Log.d(TAG, "Interstitial loaded successfully");
                    callback.onAdLoaded();
                }

                @Override
                public void onAdClicked(Ad ad) {
                    Log.d(TAG, "Interstitial clicked");
                }

                @Override
                public void onLoggingImpression(Ad ad) {
                    Log.d(TAG, "Interstitial impression logged");
                    if (currentInterstitialCallback != null) {
                        currentInterstitialCallback.onAdShown();
                    }
                }

                @Override
                public void onInterstitialDisplayed(Ad ad) {
                    Log.d(TAG, "Interstitial displayed");
                }

                @Override
                public void onInterstitialDismissed(Ad ad) {
                    Log.d(TAG, "Interstitial dismissed");
                    if (currentInterstitialCallback != null) {
                        currentInterstitialCallback.onAdClosed();
                        currentInterstitialCallback = null;
                    }
                }
            };

            interstitialAd.loadAd(interstitialAd.buildLoadAdConfig().withAdListener(interstitialAdListener).build());
        } catch (Exception e) {
            Log.e(TAG, "Error loading interstitial", e);
            callback.onError("Error loading interstitial: " + e.getMessage());
        }
    }

    public void showInterstitial(InterstitialCallback callback) {
        if (interstitialAd == null || !interstitialAd.isAdLoaded()) {
            callback.onError("Interstitial not loaded");
            return;
        }

        try {
            Log.d(TAG, "Showing interstitial");

            // Store callback for the unified listener to use
            currentInterstitialCallback = callback;

            interstitialAd.show();
        } catch (Exception e) {
            Log.e(TAG, "Error showing interstitial", e);
            currentInterstitialCallback = null;
            callback.onError("Error showing interstitial: " + e.getMessage());
        }
    }

    public boolean isInterstitialLoaded() {
        return interstitialAd != null && interstitialAd.isAdLoaded();
    }

    public void setTestMode(boolean enabled) {
        Log.d(TAG, "Setting test mode: " + enabled);
        AdSettings.setTestMode(enabled);
    }

    public void addTestDevice(String deviceId) {
        Log.d(TAG, "Adding test device: " + deviceId);
        testDevices.add(deviceId);
        if (isInitialized) {
            AdSettings.addTestDevice(deviceId);
        }
    }
}

File: ios/Sources/MetaAdsPlugin/MetaAds.swift
import Foundation
import FBAudienceNetwork

@objc public class MetaAds: NSObject {
    private var rewardedVideoAd: FBRewardedVideoAd?
    private var interstitialAd: FBInterstitialAd?
    private var isInitialized = false
    private var testDevices: [String] = []
    
    // Callback typealias
    public typealias InitializationCallback = (Bool, String?) -> Void
    public typealias AdLoadCallback = (Bool, String?) -> Void
    public typealias RewardedVideoCallback = (Bool, [String: Any]?) -> Void
    public typealias InterstitialCallback = (Bool) -> Void
    
    @objc public func initialize(appId: String, testMode: Bool, callback: @escaping InitializationCallback) {
        DispatchQueue.main.async {
            // Set test mode if enabled
            if testMode {
                FBAdSettings.setTestMode(true)
                print("MetaAds: Test mode enabled")
            }
            
            // Add test devices if any
            if !self.testDevices.isEmpty {
                FBAdSettings.addTestDevices(self.testDevices)
                print("MetaAds: Added \(self.testDevices.count) test devices")
            }
            
            // Initialize Audience Network
            FBAudienceNetworkAds.initialize(with: nil) { result in
                if result.success {
                    self.isInitialized = true
                    print("MetaAds: Meta Audience Network initialized successfully")
                    callback(true, nil)
                } else {
                    let error = result.message ?? "Unknown initialization error"
                    print("MetaAds: Failed to initialize Meta Audience Network: \(error)")
                    callback(false, "Initialization failed: \(error)")
                }
            }
        }
    }
    
    @objc public func loadRewardedVideo(placementId: String, callback: @escaping AdLoadCallback) {
        guard isInitialized else {
            callback(false, "Meta Audience Network not initialized")
            return
        }
        
        DispatchQueue.main.async {
            print("MetaAds: Loading rewarded video with placement ID: \(placementId)")
            
            self.rewardedVideoAd = FBRewardedVideoAd(placementID: placementId)
            
            guard let rewardedVideoAd = self.rewardedVideoAd else {
                callback(false, "Failed to create rewarded video ad")
                return
            }
            
            rewardedVideoAd.delegate = RewardedVideoAdDelegate(callback: callback)
            rewardedVideoAd.load()
        }
    }
    
    @objc public func showRewardedVideo(callback: @escaping RewardedVideoCallback) {
        guard let rewardedVideoAd = rewardedVideoAd, rewardedVideoAd.isAdValid else {
            callback(false, nil)
            return
        }
        
        DispatchQueue.main.async {
            print("MetaAds: Showing rewarded video")
            
            if let rootViewController = self.getRootViewController() {
                let showDelegate = RewardedVideoShowDelegate(callback: callback)
                self.rewardedVideoAd?.delegate = showDelegate
                rewardedVideoAd.show(fromRootViewController: rootViewController)
            } else {
                callback(false, nil)
            }
        }
    }
    
    @objc public func isRewardedVideoLoaded() -> Bool {
        return rewardedVideoAd?.isAdValid ?? false
    }
    
    @objc public func loadInterstitial(placementId: String, callback: @escaping AdLoadCallback) {
        guard isInitialized else {
            callback(false, "Meta Audience Network not initialized")
            return
        }
        
        DispatchQueue.main.async {
            print("MetaAds: Loading interstitial with placement ID: \(placementId)")
            
            self.interstitialAd = FBInterstitialAd(placementID: placementId)
            
            guard let interstitialAd = self.interstitialAd else {
                callback(false, "Failed to create interstitial ad")
                return
            }
            
            interstitialAd.delegate = InterstitialAdDelegate(callback: callback)
            interstitialAd.load()
        }
    }
    
    @objc public func showInterstitial(callback: @escaping InterstitialCallback) {
        guard let interstitialAd = interstitialAd, interstitialAd.isAdValid else {
            callback(false)
            return
        }
        
        DispatchQueue.main.async {
            print("MetaAds: Showing interstitial")
            
            if let rootViewController = self.getRootViewController() {
                let showDelegate = InterstitialShowDelegate(callback: callback)
                self.interstitialAd?.delegate = showDelegate
                interstitialAd.show(fromRootViewController: rootViewController)
            } else {
                callback(false)
            }
        }
    }
    
    @objc public func isInterstitialLoaded() -> Bool {
        return interstitialAd?.isAdValid ?? false
    }
    
    @objc public func setTestMode(enabled: Bool) {
        print("MetaAds: Setting test mode: \(enabled)")
        FBAdSettings.setTestMode(enabled)
    }
    
    @objc public func addTestDevice(deviceId: String) {
        print("MetaAds: Adding test device: \(deviceId)")
        testDevices.append(deviceId)
        if isInitialized {
            FBAdSettings.addTestDevice(deviceId)
        }
    }

    // MARK: - Helper Methods

    private func getRootViewController() -> UIViewController? {
        if #available(iOS 13.0, *) {
            // Use the new scene-based approach for iOS 13+
            guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                  let window = windowScene.windows.first else {
                return nil
            }
            return window.rootViewController
        } else {
            // Fallback for iOS 12 and earlier
            return UIApplication.shared.keyWindow?.rootViewController
        }
    }
}

// MARK: - Delegate Classes

private class RewardedVideoAdDelegate: NSObject, FBRewardedVideoAdDelegate {
    private let callback: MetaAds.AdLoadCallback
    
    init(callback: @escaping MetaAds.AdLoadCallback) {
        self.callback = callback
    }
    
    func rewardedVideoAdDidLoad(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video loaded successfully")
        callback(true, nil)
    }
    
    func rewardedVideoAd(_ rewardedVideoAd: FBRewardedVideoAd, didFailWithError error: Error) {
        print("MetaAds: Rewarded video failed to load: \(error.localizedDescription)")
        callback(false, "Failed to load rewarded video: \(error.localizedDescription)")
    }
    
    func rewardedVideoAdDidClick(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video clicked")
    }
    
    func rewardedVideoAdWillLogImpression(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video impression logged")
    }
}

private class RewardedVideoShowDelegate: NSObject, FBRewardedVideoAdDelegate {
    private let callback: MetaAds.RewardedVideoCallback
    
    init(callback: @escaping MetaAds.RewardedVideoCallback) {
        self.callback = callback
    }
    
    func rewardedVideoAdDidLoad(_ rewardedVideoAd: FBRewardedVideoAd) {
        // Not called during show
    }
    
    func rewardedVideoAd(_ rewardedVideoAd: FBRewardedVideoAd, didFailWithError error: Error) {
        print("MetaAds: Error showing rewarded video: \(error.localizedDescription)")
        callback(false, nil)
    }
    
    func rewardedVideoAdDidClick(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video clicked during show")
    }
    
    func rewardedVideoAdWillLogImpression(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video shown")
    }
    
    func rewardedVideoAdVideoComplete(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video completed - reward earned")
        let reward = ["type": "coins", "amount": 1] as [String: Any]
        callback(true, reward)
    }
    
    func rewardedVideoAdDidClose(_ rewardedVideoAd: FBRewardedVideoAd) {
        print("MetaAds: Rewarded video closed")
        callback(false, nil)
    }
}

private class InterstitialAdDelegate: NSObject, FBInterstitialAdDelegate {
    private let callback: MetaAds.AdLoadCallback
    
    init(callback: @escaping MetaAds.AdLoadCallback) {
        self.callback = callback
    }
    
    func interstitialAdDidLoad(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial loaded successfully")
        callback(true, nil)
    }
    
    func interstitialAd(_ interstitialAd: FBInterstitialAd, didFailWithError error: Error) {
        print("MetaAds: Interstitial failed to load: \(error.localizedDescription)")
        callback(false, "Failed to load interstitial: \(error.localizedDescription)")
    }
    
    func interstitialAdDidClick(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial clicked")
    }
    
    func interstitialAdWillLogImpression(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial impression logged")
    }
}

private class InterstitialShowDelegate: NSObject, FBInterstitialAdDelegate {
    private let callback: MetaAds.InterstitialCallback
    
    init(callback: @escaping MetaAds.InterstitialCallback) {
        self.callback = callback
    }
    
    func interstitialAdDidLoad(_ interstitialAd: FBInterstitialAd) {
        // Not called during show
    }
    
    func interstitialAd(_ interstitialAd: FBInterstitialAd, didFailWithError error: Error) {
        print("MetaAds: Error showing interstitial: \(error.localizedDescription)")
        callback(false)
    }
    
    func interstitialAdDidClick(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial clicked during show")
    }
    
    func interstitialAdWillLogImpression(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial shown")
        callback(true)
    }
    
    func interstitialAdDidClose(_ interstitialAd: FBInterstitialAd) {
        print("MetaAds: Interstitial closed")
    }
}